src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/AutoInsert/RemoteAutoInsertService.cs

```diff
@@ -0,0 +1,218 @@
+            character,
+            autoCloseTags);
+
+        if (insertTextEdit is { } edit)
+        {
+            return Response.Results(RemoteInsertTextEdit.FromLspInsertTextEdit(edit));
+        }
+
+        var positionInfo = PreferHtmlInAttributeValuesDocumentPositionInfoStrategy.Instance.GetPositionInfo(_documentMappingService, codeDocument, index);
+        var languageKind = positionInfo.LanguageKind;
+
+        if (languageKind is RazorLanguageKind.Razor)
+        {
+            // If we are in Razor and got no edit from our own service, there is nothing else to do
+            return Response.NoFurtherHandling;
+        }
+        else if (languageKind is RazorLanguageKind.Html)
+        {
+            return AutoInsertService.HtmlAllowedAutoInsertTriggerCharacters.Contains(character)
+                ? Response.CallHtml
+                : Response.NoFurtherHandling;
+        }
+        else if (languageKind is RazorLanguageKind.CSharp)
+        {
+            var mappedPosition = positionInfo.Position.ToLinePosition();
+            return await TryResolveInsertionAsyncInCSharpAsync(
+                    remoteDocumentContext,
+                    mappedPosition,
+                    character,
+                    formatOnType,
+                    indentWithTabs,
+                    indentSize,
+                    cancellationToken);
+        }
+
+        _logger.LogError($"Unsupported language {languageKind}");
+        return Response.NoFurtherHandling;
+    }
+
+    private async ValueTask<Response> TryResolveInsertionAsyncInCSharpAsync(
+        RemoteDocumentContext remoteDocumentContext,
+        LinePosition mappedPosition,
+        string character,
+        bool formatOnType,
+        bool indentWithTabs,
+        int indentSize,
+        CancellationToken cancellationToken)
+    {
+        if (!AutoInsertService.CSharpAllowedAutoInsertTriggerCharacters.Contains(character))
+        {
+            return Response.NoFurtherHandling;
+        }
+
+        // Special case for C# where we use AutoInsert for two purposes:
+        // 1. For XML documentation comments (filling out the template when typing "///")
+        // 2. For "on type formatting" style behavior, like adjusting indentation when pressing Enter inside empty braces
+        //
+        // If users have turned off on-type formatting, they don't want the behavior of number 2, but its impossible to separate
+        // that out from number 1. Typing "///" could just as easily adjust indentation on some unrelated code higher up in the
+        // file, which is exactly the behavior users complain about.
+        //
+        // Therefore we are just going to no-op if the user has turned off on type formatting. Maybe one day we can make this
+        // smarter, but at least the user can always turn the setting back on, type their "///", and turn it back off, without
+        // having to restart VS. Not the worst compromise (hopefully!)
```

Comment:
Random note: This comment is a perfect example of things we can now improve with cohosting. Since we're not using LSP to talk to Roslyn we may very well be able to separate these two concepts.

No action in this PR obv :)